src/rserver.cc

```diff
@@ -110,29 +110,46 @@ void RServerRPC::initRCore() {
 
 Status RServerRPC::FunctionCall(ServerContext *context, const CallRequest *callRequest,
                                 CallResponse *result) {
-    (void)context;
     try {
+        if (!this->mtx.try_lock()) {
+            this->rLog->log(RServerLogLevel::ERRORS,
+                            "The previous query is still running, cannot accpet new query. Maybe "
+                            "client is timeout");
+        }
         this->rLog->log(RServerLogLevel::LOGS, "start to process query");
         this->runtime->prepare(callRequest);
         this->runtime->execute();
+
+        // If cancelled, we do not need to process the results
+        if (context->IsCancelled()) {
+            this->rLog->log(RServerLogLevel::WARNINGS, "this query has been cancelled by client");
```

Comment:
unlock mutex if hold here ? The same in the catch block.

Reply:
Thx! Fixed.